### PR TITLE
*: update stats after creating table.

### PR DESCRIPTION
**Pull request by hanfei1991**
We update stats table after creating table.
@shenli @coocood @zimulala @lamxTyler PTAL

### CODE REVIEW

File: statistics/ddl.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	dropColumn
+)
+
+// DDLTask is a task describing how we should update the stats table after a ddl operation has done.
+type DDLTask struct {
+	tp        ddlType
+	tableInfo *model.TableInfo
+}
+
+// DDLCh return the ddl task channel.
```
Comment (lamxTyler) [inline statistics/ddl.go]: returns

**Comment by lamxTyler** [comment]
LGTM

**Comment by hanfei1991** [comment]
@coocood  I only updated the stats_table in TiKV. Every tidb node loads by its own.

File: statistics/ddl.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	tableInfo *model.TableInfo
+}
+
+// DDLCh returns the ddl task channel.
+func (h *Handle) DDLCh() <-chan *DDLTask {
+	return h.ddlCh
+}
+
+// CreateTable creates a ddl task and pushes to channel. A worker will read the channel and updates the stats table.
+func (h *Handle) CreateTable(tableInfo *model.TableInfo) {
```
Comment (coocood) [inline statistics/ddl.go]: This method is a notify method, should not name it `CreateTable` as if the work is done already.

File: statistics/ddl.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+// HandleDDLTask begins to process a ddl task.
+func (h *Handle) HandleDDLTask(t *DDLTask) error {
+	switch t.tp {
+	case createTable:
+		return h.onCreateTable(t.tableInfo)
+	}
+	return nil
+}
+
+func (h *Handle) onCreateTable(info *model.TableInfo) error {
```
Comment (coocood) [inline statistics/ddl.go]: This method does the actual work, the name should explicitly describe what it does.

File: domain/domain.go [hunk 4]
```diff
@@ -445,6 +445,7 @@ func (do *Domain) StatsHandle() *statistics.Handle {
 // should be called only once in BootstrapSession.
 func (do *Domain) LoadTableStatsLoop(ctx context.Context) error {
```
Comment (coocood) [inline domain/domain.go]: This method does more that the method indicates now, should be renamed to contain the added work.

File: domain/domain.go [hunk 5]
```diff
@@ -466,6 +467,11 @@ func (do *Domain) LoadTableStatsLoop(ctx context.Context) error {
 				}
 			case <-do.exit:
 				return
+			case t := <-do.statsHandle.DDLCh():
```
Comment (coocood) [inline domain/domain.go]: Add comments about this channel is sent by DDL owner.

File: statistics/statscache.go [hunk 6]
```diff
@@ -43,7 +44,8 @@ func (h *Handle) Clear() {
 // NewHandle creates a Handle for update stats.
 func NewHandle(ctx context.Context) *Handle {
 	handle := &Handle{
-		ctx: ctx,
+		ctx:   ctx,
+		ddlCh: make(chan *DDLTask, 10000),
```
Comment (coocood) [inline statistics/statscache.go]: Why make it so large?

Comment (hanfei1991) [inline statistics/statscache.go]: How much is ok ?

Comment (coocood) [inline statistics/statscache.go]: 100 is large enough

File: statistics/ddl.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+const (
+	createTable ddlType = iota
+	dropTable
+	createColumn
+	dropColumn
+)
+
+// DDLTask is a task describing how we should update the stats table after a ddl operation has done.
+type DDLTask struct {
```
Comment (coocood) [inline statistics/ddl.go]: DDLEvent is better

**Comment by hanfei1991** [comment]
@coocood @zimulala PTAL

File: statistics/ddl.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+	return h.ddlCh
+}
+
+// NotifyCreateTable creates a ddl task and pushes to channel. A worker will read the channel and updates the stats table.
+func (h *Handle) NotifyCreateTable(tableInfo *model.TableInfo) {
+	h.ddlCh <- &DDLEvent{tp: createTable, tableInfo: tableInfo}
+}
+
+// HandleDDLTask begins to process a ddl task.
+func (h *Handle) HandleDDLTask(t *DDLEvent) error {
```
Comment (coocood) [inline statistics/ddl.go]: HandleDDLEvent

**Comment by coocood** [comment]
ddl package can only keep a channel that recieves ddl event, it doesn't need to depend statistics package.

File: statistics/statscache.go [hunk 9]
```diff
@@ -32,6 +33,9 @@ type Handle struct {
 	ctx         context.Context
 	lastVersion uint64
 	statsCache  atomic.Value
+	// ddlEventCh is a channel to notify a ddl operation has happened. It is sent only by owner and read by stats handle.
+	// Exported for test.
```
Comment (coocood) [inline statistics/statscache.go]: Not exported.

**Pull request merged**